### PR TITLE
[Coroutines] Introduce [[clang::coro_only_destroy_when_complete]]

### LINKED ISSUE #56980: Feature request: allow teaching llvm that coroutines can only be destroyed when done
The following code contains two things:

* A basic coroutine task type that can be awaited from other coroutines returning that task type. It is not possible to do anything with the task except await it. In particular it's not possible to destroy the task's coroutine early: **coroutines are always destroyed only when they are done**.

* A coroutine called `Foo` returning that task type that has several different suspend points, each with a different instance of `GlobalSetter` alive. When `GlobalSetter` is destroyed, it sets `some_global` in a way that can be seen in assembly code.

```c++
#include <coroutine>

class MyTask{
 public:
  struct promise_type {
    MyTask get_return_object() { return MyTask(std::coroutine_handle<promise_type>::from_promise(*this)); }
    std::suspend_always initial_suspend() { return {}; }

    void unhandled_exception();

    void return_void() {} 

    auto await_transform(MyTask task) {
      struct Awaiter {
        bool await_ready() { return false; }
        std::coroutine_handle<promise_type> await_suspend(std::coroutine_handle<promise_type> h) {
          caller.resume_when_done = h;
          return std::coroutine_handle<promise_type>::from_promise(callee);
        }

        void await_resume() {
          std::coroutine_handle<promise_type>::from_promise(callee).destroy();
        }

        promise_type& caller;
        promise_type& callee;
      };

      return Awaiter{*this, task.handle_.promise()};
    }
    
    auto final_suspend() noexcept {
      struct Awaiter {
        bool await_ready() noexcept { return false; }
        std::coroutine_handle<promise_type> await_suspend(std::coroutine_handle<promise_type> h) noexcept {
          return to_resume;
        }

        void await_resume() noexcept;

        std::coroutine_handle<promise_type> to_resume;
      };

      return Awaiter{resume_when_done};
    }

    // The coroutine to resume when we're done.
    std::coroutine_handle<promise_type> resume_when_done;
  };


 private:
  explicit MyTask(const std::coroutine_handle<promise_type> handle)
      : handle_(handle) {}

  // A handle for the coroutine that returned this task. This is not
  // revealed to the user: it can only be used to await the task.
  const std::coroutine_handle<promise_type> handle_;
};

int some_global;

// A type whose destructor takes an action that's easy to see in the output
// assembly code.
struct GlobalSetter {
  int value;

  GlobalSetter(int value) : value(value) {}
  ~GlobalSetter() { some_global = value; }
};

MyTask __attribute__((noinline)) Qux() { co_return; }

MyTask Foo() {
  {
    GlobalSetter gs(0);
    co_await Qux();
  }

  {
    GlobalSetter gs(1);
    co_await Qux();
  }

  {
    GlobalSetter gs(2);
    co_await Qux();
  }

  {
    GlobalSetter gs(3);
    co_await Qux();
  }

  {
    GlobalSetter gs(4);
    co_await Qux();
  }

  {
    GlobalSetter gs(5);
    co_await Qux();
  }

  {
    GlobalSetter gs(6);
    co_await Qux();
  }
}
```

Because it's not possible to destroy the frame for `Foo` until it is completely finished, the cleanup code in theory can be trivial: there is no need to worry about running `~GlobalSetter`, because that will always have already run in the coroutine body itself. But when compiled with `-std=c++20 -O2 -fno-exceptions`, clang is unable to see this and generates a bunch of dead code within `Foo.destroy` ([compiler explorer](https://godbolt.org/z/9K8eKE9Kh)):

```asm
Foo() [clone .destroy]:                        # @Foo() [clone .destroy]
        cmp     qword ptr [rdi], 0
        je      .LBB5_9
        movzx   ecx, byte ptr [rdi + 96]
        xor     eax, eax
        lea     rdx, [rip + .LJTI5_0]
        movsxd  rcx, dword ptr [rdx + 4*rcx]
        add     rcx, rdx
        jmp     rcx
.LBB5_2:
        mov     eax, 1
        jmp     .LBB5_8
.LBB5_3:
        mov     eax, 2
        jmp     .LBB5_8
.LBB5_4:
        mov     eax, 3
        jmp     .LBB5_8
.LBB5_5:
        mov     eax, 4
        jmp     .LBB5_8
.LBB5_6:
        mov     eax, 5
        jmp     .LBB5_8
.LBB5_7:
        mov     eax, 6
.LBB5_8:
        mov     dword ptr [rip + some_global], eax
.LBB5_9:
        jmp     operator delete(void*)@PLT                      # TAILCALL
.LJTI5_0:
        .long   .LBB5_9-.LJTI5_0
        .long   .LBB5_8-.LJTI5_0
        .long   .LBB5_2-.LJTI5_0
        .long   .LBB5_3-.LJTI5_0
        .long   .LBB5_4-.LJTI5_0
        .long   .LBB5_5-.LJTI5_0
        .long   .LBB5_6-.LJTI5_0
        .long   .LBB5_7-.LJTI5_0
```

In theory **this should just be a single instruction**: a tail call to `operator delete`. Instead we get generated code that duplicates all of the logic for destroying the `GlobalSetter` objects that is also in `Foo.resume`.

This is a feature request to **add some facility to teach llvm that generating code for destroying local variables is unnecessary in the coroutine destroy function**. I thought perhaps adding a `__builtin_unreachable` section in the promise's destructor that you hit when the promise's handle is not done might work, but [no such luck](https://godbolt.org/z/jf5K78hn8).

### PR DESCRIPTION
**Pull request by ChuanqiXu9**
Close https://github.com/llvm/llvm-project/issues/56980.

This patch tries to introduce a light-weight optimization attribute for coroutines which are guaranteed to only be destroyed after it reached the final suspend.

The rationale behind the patch is simple. See the example:

```C++
A foo() {
  dtor d;
  co_await something();
  dtor d1;
  co_await something();
  dtor d2;
  co_return 43;
}
```

Generally the generated .destroy function may be:

```C++
void foo.destroy(foo.Frame *frame) {
  switch(frame->suspend_index()) {
    case 1:
      frame->d.~dtor();
      break;
    case 2:
      frame->d.~dtor();
      frame->d1.~dtor();
      break;
    case 3:
      frame->d.~dtor();
      frame->d1.~dtor();
      frame->d2.~dtor();
      break;
    default: // coroutine completed or haven't started
      break;
  }

  frame->promise.~promise_type();
  delete frame;
}
```

Since the compiler need to be ready for all the cases that the coroutine may be destroyed in a valid state.

However, from the user's perspective, we can understand that certain coroutine types may only be destroyed after it reached to the final suspend point. And we need a method to teach the compiler about this. Then this is the patch. After the compiler recognized that the coroutines can only be destroyed after complete, it can optimize the above example to:

```C++
void foo.destroy(foo.Frame *frame) {
  frame->promise.~promise_type();
  delete frame;
}
```

I spent a lot of time experimenting and experiencing this in the downstream. The numbers are really good. In a real-world coroutine-heavy workload, the size of the build dir (including .o files) reduces 14%. And the size of final libraries (excluding the .o files) reduces 8% in Debug mode and 1% in Release mode.

### CODE REVIEW

File: clang/lib/CodeGen/CGCoroutine.cpp
Comment (tbaederr) [inline clang/lib/CodeGen/CGCoroutine.cpp:785]:
```suggestion
  if (const CXXRecordDecl *RD = FnRetTy->getAsCXXRecordDecl();
      RD && RD->hasAttr<CoroOnlyDestroyWhenCompleteAttr>())
    CurFn->setCoroDestroyOnlyWhenDone();
```

Comment (ChuanqiXu9) [reply clang/lib/CodeGen/CGCoroutine.cpp:785]: Got it. Thanks. Will do in the next round or landing.

File: clang/include/clang/Basic/AttrDocs.td
Comment (erichkeane) [inline clang/include/clang/Basic/AttrDocs.td:7467]: For the section below, how about:

```
The `foo.destroy()` function's purpose is to release all of the resources initialized for the coroutine when it is destroyed in a suspended state.  However, if the coroutine is only ever destroyed at the final suspend state, the rest of the conditions are superfluous. The user can use the `coro_only_destroy_when_complete` attributo suppress generation of the other destruction cases, optimizing the above `foo.destroy` to:
```

**Review by erichkeane — COMMENTED** [review]
Didn't dig into the LLVM section of this, but the CFE section looks good delta some small documentation changes.

**Review by erichkeane — APPROVED** [review]
CFE stuff looks good, but someone more familiar with the LLVM changes needs to do final approval.

File: clang/include/clang/Basic/AttrDocs.td
Comment (ChuanqiXu9) [reply clang/include/clang/Basic/AttrDocs.td]: I changed this to `This is helpful for the optimizers to reduce the size of the destroy function for the coroutines`.

To be honest, I didn't think this is a problem. For example, [[likey]], [[unlikely]] and [[assume]] don't explain what optimization would be performed actually. The contract between the users and the compilers is: the users should provide the correct information then the compiler at least shouldn't generate worse code.

BTW, an interesting point here is that, according to the design intention of coroutines, users shouldn't care about what is `destroy` function of coroutines. But actually, many C++ programmers (who don't know about compilers) understand the concept of `destroy` functions. This may be the culture of C++ : )

Comment (erichkeane) [reply clang/include/clang/Basic/AttrDocs.td]: Its better enough to be an 'intro' sentence at least.  And yes, our other documentation is bad, but we're trying to improve it.

Those are at least better 'understood' (or at least, understood in their lack of understanding) attributes that most folks can reason about, since they are documented plenty elsewhere (blogs, standards, etc).

**Comment by erichkeane** [comment]
Also note: This needs a Release Note in ReleaseNotes.rst.

**Review by erichkeane — APPROVED** [review]
CFE + Release note LGTM.  Again, please wait on someone to be review the LLVM changes.

File: clang/include/clang/Basic/AttrDocs.td
Comment (ChuanqiXu9) [reply clang/include/clang/Basic/AttrDocs.td]: Oh, I am not saying other documentation is bad. I mean, from the perspective of optimizers, it will never be able (or better not to) explain what the optimizations made by [[likey]], [[unlikely]] and [[assume]]. Since the optimizers want the freedom to perform any optimization they want as long as they doesn't violate the semantics. Otherwise, it'll be the responsibility of optimizers to report what they did after they made some non-trivial changes. It is generally not wanted by optimizers. We wish the programmers and implementers can focus more on semantics.

Comment (erichkeane) [reply clang/include/clang/Basic/AttrDocs.td]: Right, I'm just suggesting we give 'motivating' text for folks to read, not "this pass does X, this pass does Y". In this case, if I were to see "better optimizations" that doesn't help/isn't particularly actionable.  But if I see, "Allows for better code size in cases where a coroutine is known to ONLY be destroyed after its final suspend point", and I have one of those, it might be 'relevant to my interests!'.

While the example was good, I found I wasn't able to determine WHAT the attribute was doing until investing a lot of time reading, we want our users/doc readers to be 'hooked' on being interest (or, alternatively, be able to tell they don't need to continue reading) as soon as possible.  Does that make sense?

Comment (ChuanqiXu9) [reply clang/include/clang/Basic/AttrDocs.td]: Got. Completely understood. Many thanks for reviewing this and sharing thoughts : )

**Review by h-vetinari — COMMENTED** [review]
There is substantial inconsistency between "coro only destroy when complete" and "... when done" throughout the PR (in various forms; e.g. literal strings, snake_case, CamelCase, ALL_CAPS as well as file names).

Presumably there should only be one spelling everywhere.

**Comment by ChuanqiXu9** [comment]
> There is substantial inconsistency between "coro only destroy when complete" and "... when done" throughout the PR (in various forms; e.g. literal strings, snake_case, CamelCase, ALL_CAPS as well as file names).
> 
> Presumably there should only be one spelling everywhere.

Done

**Comment by ChuanqiXu9** [comment]
@bcardosolopes @apolloww @rjmccall would you like to take a look for the middle end part?

File: clang/test/CodeGenCoroutines/coro-only-destroy-when-complete.cpp
Comment (h-vetinari) [inline clang/test/CodeGenCoroutines/coro-only-destroy-when-complete.cpp:54]: still `_done` not `_complete`

File: llvm/docs/Coroutines.rst
Comment (h-vetinari) [inline llvm/docs/Coroutines.rst:1784]: still `_done` not `_complete`

File: llvm/include/llvm/Bitcode/LLVMBitCodes.h
Comment (h-vetinari) [inline llvm/include/llvm/Bitcode/LLVMBitCodes.h:721]: still `_DONE` not `_COMPLETE` (plus all later usages of it).

File: llvm/include/llvm/IR/Attributes.td
Comment (h-vetinari) [inline llvm/include/llvm/IR/Attributes.td:322]: still `_done` not `_complete`

File: llvm/test/Transforms/Coroutines/coro-only-destroy-when-done.ll
Comment (h-vetinari) [inline llvm/test/Transforms/Coroutines/coro-only-destroy-when-done.ll:1]: filename still has `-done` not `-complete`

File: llvm/test/Transforms/Coroutines/coro-only-destroy-when-done.ll
Comment (h-vetinari) [inline llvm/test/Transforms/Coroutines/coro-only-destroy-when-done.ll:117]: still `_done` not `_complete`

**Review by h-vetinari — COMMENTED** [review]
> > Presumably there should only be one spelling everywhere.
>
> Done

Missed a few. ;-)

File: clang/test/CodeGenCoroutines/coro-only-destroy-when-complete.cpp
Comment (ChuanqiXu9) [reply clang/test/CodeGenCoroutines/coro-only-destroy-when-complete.cpp:35]: > Since the attribute is on the return type, would that mean it would apply to any coroutine returning the particular type?

Yes.

> Say user is using some library implementation (i.e. folly::coro::Task), they would not be able to use the attribute because it would mark all the coroutines this way.

Yes. My thought is that such attribute can only be applied if the programmers understand the behavior of the coroutines very well. So that the users of the attribute should be the library writers instead of end users.

And for `folly::coro::Task`, (I didn't check it actively), it can't use the attribute if I remember correctly since its framework supports cancelling, that said, a task can be cancelled before being complete.

Comment (ChuanqiXu9) [reply clang/test/CodeGenCoroutines/coro-only-destroy-when-complete.cpp:35]: BTW, while it is technically possible to give finer grained marks, I don't think it is good. Since the behaviors of coroutines are controlled by the framework, in another word, the library writers. And it is not good for the end users to predict the behaviors.

**Comment by ChuanqiXu9** [comment]
> > > Presumably there should only be one spelling everywhere.
> > 
> > 
> > Done
> 
> Missed a few. ;-)

Sorry and thanks for the detailed comment!

File: clang/test/CodeGenCoroutines/coro-only-destroy-when-complete.cpp
Comment (apolloww) [reply clang/test/CodeGenCoroutines/coro-only-destroy-when-complete.cpp:35]: Thanks for the explaination

**Review by apolloww — APPROVED** [review]

**Comment by ChuanqiXu9** [comment]
Thanks for reviewing this!

**Pull request merged**
